### PR TITLE
[router] Impl radix tree and set up CI

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

1. implement a radix tree for later use
2. set up CI for rust

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: .github/workflows/pr-test-rust.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    paths:
+      - "rust/*"
+  pull_request:
+    branches: [ main ]
+    paths:
+      - "rust/*"
+  workflow_dispatch:
+
+concurrency:
+  group: pr-test-${{ github.ref }}
```
Comment (merrymercy) [inline .github/workflows/pr-test-rust.yml]: Use a different name to resolve conflict. Otherwise, it will cancel other workflows.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-test-rust.yml (hunk 1 above)
Comment (ByronHsu) [reply .github/workflows/pr-test-rust.yml]: fixed

File: rust/src/tree.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
+#[derive(Clone)]
+pub struct Node {
+    pub children: Vec<Node>,
```
Comment (Ying1123) [inline rust/src/tree.rs]: Use dict of [first char] -> Node, rather than vector.

File: .github/workflows/pr-test-rust.yml (hunk 1 above)
Comment (ByronHsu) [reply .github/workflows/pr-test-rust.yml]: @merrymercy do you know what is the pending "finish" in the ci?

File: rust/src/tree.rs
Comment (Ying1123) [inline rust/src/tree.rs:1]: It's fine as a vanilla version for now. In the future, we'll need to support multiple processes/threads to access the tree. So race condition should be avoided (e.x. using Arc and RwLock).

File: rust/src/tree.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                // if not full match, break
+                break;
+            }
+        }
+
+        return &input_ids[..curr_idx];
+
+    }
+
+    pub fn delete(&mut self, input_ids: &[usize]) {
```
Comment (Ying1123) [inline rust/src/tree.rs:124]: In the future PRs, we might want to support delete with node contractions.

**Review by Ying1123 — APPROVED** [review]

File: rust/src/tree.rs (hunk 2 above)
Comment (ByronHsu) [reply rust/src/tree.rs]: Made the change. great suggestion

File: rust/src/tree.rs
Comment (ByronHsu) [reply rust/src/tree.rs:1]: Will do in the follow-up PRs!

File: rust/src/tree.rs (hunk 3 above)
Comment (ByronHsu) [reply rust/src/tree.rs:124]: makes sense

**Comment by ByronHsu** [comment]
@Ying1123 I have 
1. Fixed the `clone()` overhead by `mem::take`
2. Replaced vector with hashmap
3. Added fmt check in ci

Thanks for the great feedback

File: rust/src/tree.rs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        let mut curr = &mut self.root;
+        curr.count += 1;
+
+        let mut curr_idx = 0;
+        let input_ids_len = input_ids.len();
+
+        while curr_idx < input_ids_len {
+            let mut match_info = None;
+
+            for key in curr.children.keys() {
```
Comment (Ying1123) [inline rust/src/tree.rs]: Since you have changed the children to be a hashmap. The for loop is not needed here.

File: rust/src/tree.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    pub fn prefix_match<'a>(&self, input_ids: &'a [usize]) -> &'a [usize] {
+        let mut curr = &self.root;
+
+        let mut curr_idx = 0;
+        let input_ids_len = input_ids.len();
+
+        while curr_idx < input_ids_len {
+            let mut has_full_match = false;
+
+            for key in curr.children.keys() {
```
Comment (Ying1123) [inline rust/src/tree.rs]: Same here

File: rust/src/tree.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+        let mut curr_idx = 0;
+        let input_ids_len = input_ids.len();
+
+        while curr_idx < input_ids_len {
+            // First find the matching child index and prefix length
+            let mut child_key = None;
+            let mut prefix_len = 0;
+
+            for key in curr.children.keys() {
```
Comment (Ying1123) [inline rust/src/tree.rs]: Same here

**Review by Ying1123 — APPROVED** [review]

File: .github/workflows/pr-test-rust.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+name: PR Test
```
Comment (merrymercy) [inline .github/workflows/pr-test-rust.yml]:
```suggestion
name: PR Test (Rust)
```

**Review by merrymercy — APPROVED** [review]

**Pull request merged**